wallet/src/lib.rs

```diff
@@ -0,0 +1,296 @@
+#[contracttype]
+pub struct WeightedSigners {
+    pub signers: Vec<Identifier>,
+    pub weight: u32,
+}
+
+#[derive(Clone)]
+#[contracttype]
+pub struct Admin {
+    pub id: Identifier,
+    pub weight: u32,
+}
+
+const MAX_ADMINS: u32 = 20;
+const MAX_WEIGHT: u32 = 100;
+
+pub struct WalletContract;
+
+// Contract usage:
+// - Call `initialize` once to setup the contract admins, their weights and
+//   payment weight threshold. For simplicity, this setup is immutable.
+// - Fund the wallet contract as needed using token contract functionality.
+// - To execute the payment:
+//   1. Distribute a pair of `(payment_id, Payment)` to the wallet admins for
+//      signing. `payment_id` should be unique for every payment. `payment_id`
+//      management is not implemented here for the sake of conciseness and could
+//      happen both off-chain or in the contract itself.
+//   2. Call `pay` one or many times with arbitrary batches of admin signatures
+//      until enough admin weight accumulated (i.e. at least `threshold`) to
+//      actually execute it.
+#[contractimpl]
+impl WalletContract {
+    // Performs contract intialization.
+    // Call `initialize` and supply ids and weights of the admins, as well as the
+    // threshold needed to execute payments. The payment may only be executed when
+    // unique admins with combined weight exceeding `threshold` have signed it.
+    pub fn initialize(env: Env, admins: Vec<Admin>, threshold: u32) {
+        check_initialization_params(&env, &admins, threshold);
+
+        let mut weight_sum = 0;
+        for maybe_admin in admins.iter() {
+            let admin = maybe_admin.unwrap();
+            if admin.weight == 0 {
+                panic!("weight should be non-zero");
+            }
+            if admin.weight > MAX_WEIGHT {
+                panic!("too high admin weight");
+            }
+            weight_sum += admin.weight;
+            // Record admin weight (and effectively admin identifier too).
+            env.contract_data()
+                .set(DataKey::AdminW(admin.id), admin.weight);
+        }
+        // Do a basic sanity check to make sure we don't create a locked wallet.
+        if weight_sum < threshold {
+            panic!("admin weight is lower than threshold");
+        }
+        env.contract_data().set(DataKey::Threshold, threshold);
+    }
+
+    // Helper to get nonce for any provided admin identifier.
+    // The nonce should then be used in `pay`.
+    pub fn get_nonce(env: Env, admin: Identifier) -> BigInt {
+        read_nonce(&env, admin)
+    }
+
+    // Stores a provided payment or executes it when enough signer weight is
+    // accumulated.
+    // Returns `true` when the payment was executed and `false` otherwise.
+    //
+    // Every wallet admin signs `pay` as if it was called by them only, i.e.
+    // they should sign `pay` function call with argument tuple of
+    // `(id, nonce, payment_id, payment)`. Then the signatures and nonces of
```

Comment:
We should encourage including a domain separator in anything that gets signed, so we should probably add it to this example, if we can find a concise way to do so.

Reply:
I'm not sure what do you mean by the domain separator here. Network id is a part of the signature payload expected by `verify`. Do you mean something else here?

Reply:
Ok, I missed that this section was using soroban-auth. As long as it's using the signature payload from that crate then it's good.